Convert apiquery to backfill

piuccio opened the pull request:
Backfill is now an object with type and query. This allows future extensions to queries that are not backed by capi.

As a first step both `apiQuery` and `backfill` are available, the client tries to use `backfill` first and `apiQuery` if missing.

There's a change in a public method interface so I've created a new method with a different signature, users of the library have time to migrate.

@janua @DiegoVazquezNanini @davidfurey

piuccio commented:
I've update this PR, it now has a `BackfillResover` that implements two types of backfill query

* `capi`: the traditional capi search string
* `collection`: inherit the **curated content** of a parent collection

I don't think the child collection should inherit the parent's backfill, at least until we solve loops and deduping.

@janua I've added unit tests!

janua commented on fapi-client/src/main/scala/com/gu/facia/api/utils/Backfill.scala:
```diff
@@ -0,0 +1,63 @@
+object BackfillContent {
+  def resolveFromConfig(collectionConfig: CollectionConfig): BackfillResolver = {
+    collectionConfig.backfill match {
+      case Some(backfill: Backfill) => backfill.`type` match {
```
Switch to nested case matching here.

janua commented on fapi-client/src/main/scala/com/gu/facia/api/utils/Backfill.scala:
```diff
@@ -0,0 +1,63 @@
+  }
+}
+
+case class EmptyBackfill() extends BackfillResolver {
```
This can be an `object`.

janua commented on fapi-client/src/main/scala/com/gu/facia/api/utils/Backfill.scala:
```diff
@@ -0,0 +1,63 @@
+      case Some(backfill: Backfill) => backfill.`type` match {
+        case "capi" => CapiBackfill(backfill.query, collectionConfig)
+        case "collection" => CollectionBackfill(backfill.query)
+        case _ => throw new Error(s"Invalid backfill type ${backfill.`type`}")
```
Are you sure you want an `Exception` here? Why not `EmptyBackfill`?
If so, write a quick unit test for it to show that's what is expected.

piuccio commented:
thanks @janua I've fixed your comments.

I prefer to throw an exception when the backfill is invalid because it's a configuration error, it's something that shouldn't happen. Returning an empty list will silently ignore an invalid state. I've added a test for that

piuccio commented:
@janua I've updated the PR with a better (???) backfill resolver

janua commented on fapi-client/src/main/scala/com/gu/facia/api/utils/Backfill.scala:
```diff
@@ -0,0 +1,55 @@
+case class InvalidBackfillConfiguration(msg: String) extends Error(msg)
+
+object BackfillContent {
+  def resolveFromConfig(collectionConfig: CollectionConfig): BackfillResolver = {
```
This could now go into `BackfillResolver`!

Is the name `BackfillContent` better than `BackfillResolver`?

piuccio replied:
they are now in the same companion object

janua commented on fapi-client/src/main/scala/com/gu/facia/api/utils/Backfill.scala:
```diff
@@ -0,0 +1,53 @@
+
+import scala.concurrent.ExecutionContext
+
+case class InvalidBackfillConfiguration(msg: String) extends Error(msg)
```
Don't extend `Error`. Instead extend `Exception`.

`Error` is used for critical errors such as `ThreadDeath` and `VirtualMachineError`.

piuccio replied:
fixed :smile:

janua commented on fapi-client/src/main/scala/com/gu/facia/api/FAPI.scala:
```diff
@@ -170,4 +172,17 @@ object FAPI {
+                         adjustSearchQuery: AdjustSearchQuery = identity, adjustItemQuery: AdjustItemQuery = identity)
+                        (implicit capiClient: GuardianContentClient, faciaClient: ApiClient, ec: ExecutionContext): Response[List[FaciaContent]] = {
+
+    val resolver = BackfillResolver.resolveFromConfig(collection.collectionConfig)
```
`backfillRequest`? `resolver` means nothing to me for this `val`!

piuccio replied:
fixed :smile:

janua commented:
Looks good :clap: :+1:
